After updating httpx to the latest version, I’m consistently seeing a panic when httpx tries to classify certain HTML pages (likely deeply nested or malformed HTML).

Panic message:

panic: html: open stack of elements exceeds 512 nodes

Stack trace (from httpx):
github.com/projectdiscovery/httpx/common/pagetypeclassifier.htmlToText
github.com/projectdiscovery/httpx/common/pagetypeclassifier.(*PageTypeClassifier).Classify
github.com/projectdiscovery/httpx/runner.(*Runner).analyze
github.com/projectdiscovery/httpx/runner.(*Runner).process.func1
...

I am unsure of what triggered this but basically I was using httpx like this:
"cat hosts.txt| httpx -cl -sc -title -td -ct -server -vhost -o httpx-out.txt",

Now I have updated httpx on my 2 vps machine and both when doing this exhaustive lists of hosts, errored out like this so I do not know the actual url's that triggered this error. But in my localmachine where my httpx is still the older version, it worked completely fine.


I believe there is an issue in this new update, I am very bad at this hence I am really sorry if i have wasted your time in creating this issue if it is not legitimate, but this is what I noticed like after using httpx today morning and after updating it to the latest version.

<img width="1909" height="263" alt="Image" src="https://github.com/user-attachments/assets/1cb4b8a4-4ff1-4624-9ced-4bfa81b3c183" />

--- a/common/pagetypeclassifier/pagetypeclassifier.go
+++ b/common/pagetypeclassifier/pagetypeclassifier.go
@@ -4,10 +4,11 @@ import (
 	_ "embed"
 	"sync"
 
-	"github.com/microcosm-cc/bluemonday"
-	htmltomarkdown "github.com/JohannesKaufmann/html-to-markdown/v2"
-	"github.com/projectdiscovery/utils/ml/naive_bayes"
 	"fmt"
+
+	htmltomarkdown "github.com/JohannesKaufmann/html-to-markdown/v2"
+	"github.com/microcosm-cc/bluemonday"
+	"github.com/projectdiscovery/utils/ml/naive_bayes"
 )
 
 //go:embed clf.gob
@@ -63,7 +64,7 @@ func getSanitizerPolicy() *bluemonday.Policy {
 // Strategy:
 // 1. Always sanitize HTML with bluemonday first to remove useless elements and reduce nesting
 // 2. Convert sanitized HTML to markdown
-// 3. If conversion panics, recover and return empty string
+// 3. If conversion panics, recover and return empty string with error
 func htmlToText(html string) (text string, err error) {
 	defer func() {
 		if r := recover(); r != nil {
@@ -71,20 +72,20 @@ func htmlToText(html string) (text string, err error) {
 			text = ""
 		}
 	}()
-	
+
 	// First, sanitize HTML with bluemonday to strip useless elements and reduce nesting
 	sanitizedHTML := getSanitizerPolicy().Sanitize(html)
-	
+
 	// If sanitization failed or produced empty result, return empty
 	if sanitizedHTML == "" {
 		return "", nil
 	}
-	
+
 	// Convert sanitized HTML to markdown
 	text, err = htmltomarkdown.ConvertString(sanitizedHTML)
 	if err != nil || text == "" {
 		return "", err
 	}
-	
+
 	return
 }

--- a/common/pagetypeclassifier/pagetypeclassifier_test.go
+++ b/common/pagetypeclassifier/pagetypeclassifier_test.go
@@ -90,9 +90,9 @@ func TestPageTypeClassifier(t *testing.T) {
 		}
 		deeplyNestedHTML += "</div>"
 
-		// Should not panic and should return empty string on panic
+		// Should not panic and should return empty string with error on panic
 		result, err := htmlToText(deeplyNestedHTML)
-		require.NoError(t, err)
+		require.Error(t, err)
 		require.Equal(t, "", result)
 	})
 

